core/src/main/java/org/apache/iceberg/io/inmemory/InMemoryCatalog.java

```diff
@@ -0,0 +1,374 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io.inmemory;
```

Comment:
Nit: Maybe the catalog itself deserves its own package? Everything in `org.apache.iceberg.io` (at least from `core`) seems to be just IO related.

I'd personally put the actual FileIO stuff directly into `org.apache.iceberg.io`, as most of it begins with `InMemory` and there are a number of files in that package already. You could also do `org.apache.iceberg.io.memory` if you wanted, but I don't feel the extra level is needed based on the current contents.

Then for the in-memory catalog implementation, I'd add it directly to `org.apache.iceberg.catalog`, but put it in `core` not `api`.

But the catalog itself should probably not be in package `io` like this, as everything there is concerned with readers / writers and not actual catalog specific stuff.

Just my 2 cents 😄